File: Trees/Python/kth_smallest_elem_BST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.counter = 1
        self.res = -1

        def inorder(root):
            if not root or self.res > 0:
                return
            inorder(root.left)
            if self.counter == k:
                self.res = root.val
            self.counter += 1
            inorder(root.right)

        inorder(root)
        return self.res

File: Trees/C_Cpp/kth_smallest_elem_BST.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


int kthSmallest(struct TreeNode* root, int k) {
    struct TreeNode* stack[1000];
    struct TreeNode* cur = root;
    int sp = 0;
    int counter = 1;

    while(!cur || sp > 1) {
        while(cur) {
            stack[sp++] = cur;
            cur = cur->left;
        }
        cur = stack[--sp];
        if (counter == k) {
            return cur->val;
        }
        counter++;
        cur = cur->right;
    }
    return -1;
}
